### PR TITLE
wasi: add documentation

### PR DESCRIPTION
**Pull request by cjihrig**

### CODE REVIEW

File: doc/api/wasi.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# WebAssembly System Interface (WASI)
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+The WASI API provides an implementation of the [WebAssembly System Interface][]
```
Comment (benjamingr) [inline doc/api/wasi.md:7]: I think an explanation that WASI is basically a thing that lets us use file and socket I/O from WASM in Node.js might be useful here.

Most of our users probably aren't even sure what POSIX or system calls are.

Comment (cjihrig) [reply doc/api/wasi.md:7]: @benjamingr any chance you could provide the specific text you'd like to see? You can leave it as a comment or push directly to the branch as you see fit.

**Review by guybedford — APPROVED** [review]
Seems very clear to me. A complete end-to-end example of loading a real WASI command might be a nice to have for those who don't want to strain to fill in the gaps themselves.

**Comment by cjihrig** [comment]
@guybedford do you mean the line `const wasm = getWebAssemblyModuleSomehow();`? That's the only part that isn't "real." I did that because there are a number of ways to "get" a wasm module, and that's how our streams docs do things. Do you have a preference on what you'd like to see there instead?

**Comment by guybedford** [comment]
@cjihrig I was just thinking for users who aren't familiar with WASI if they were curious and this were their first introduction of playing around with it, a very simple:

```js
const module = await WebAssembly.compile(fs.readFileSync('./binary.wasm'));
```

might just help.

Btw what is the status of stdout handling? Is it possible to setup piping or does it default to the same stdout as the Node.js process still?

Edit: removed the unnecessary base64 from my copy-paste example.

**Comment by cjihrig** [comment]
> Is it possible to setup piping or does it default to the same stdout as the Node.js process still?

[stdio currently uses the real file descriptors](https://github.com/cjihrig/uvwasi/blob/41e39d457a6158c7af4d4d490c35fe496b914932/src/fd_table.c#L255-L281). However, I *think* it would be pretty straightforward to pass alternative file descriptors to `uvwasi_init()` and use those instead.

**Comment by cjihrig** [comment]
I've added @guybedford's suggested line of code to the example.

**Comment by guybedford** [comment]
👍 

> stdio currently uses the real file descriptors. However, I think it would be pretty straightforward to pass alternative file descriptors to uvwasi_init() and use those instead.

It would be great to see an approach here further, as this is important to the command virtualization.

**Review by danbev — APPROVED** [review]

File: doc/api/wasi.md (hunk 1 above)
Comment (cjihrig) [reply doc/api/wasi.md:7]: ping @benjamingr

**Comment by cjihrig** [comment]
@guybedford I've added the file descriptor mapping piece to the checklist in https://github.com/nodejs/wasi/issues/21 so it doesn't get forgotten about.

**Review by saghul — APPROVED** [review]

**Comment by cjihrig** [comment]
Landed in https://github.com/nodejs/wasi/commit/41a236085cca7b9015606a51eeeb58085c94e6d2. Thanks for the reviews!

@benjamingr please open a PR for any docs changes/additions you'd like to see.

**Pull request closed without merging**